Index: battle.c
===================================================================
--- battle.c	(revision 15066)
+++ battle.c	(working copy)
@@ -3458,6 +3475,14 @@
 			struct map_session_data *sd = BL_CAST(BL_PC, s_bl);
 			if( s_bl != t_bl )
 			{
+				if(map[src->m].flag.team && sd->state.team > 0 && sd->state.team == ((TBL_PC*)t_bl)->state.team)
+					return 0;
+				
 				if( sd->state.killer )
 				{
 					state |= BCT_ENEMY; // Can kill anything
Index: map.h
===================================================================
--- map.h	(revision 15066)
+++ map.h	(working copy)
@@ -492,6 +492,10 @@
 		unsigned guildlock :1;
 		unsigned src4instance : 1; // To flag this map when it's used as a src map for instances
 		unsigned reset :1; // [Daegaladh]
+		unsigned team :1;
 	} flag;
 	struct point save;
 	struct npc_data *npc[MAX_NPC_PER_MAP];
Index: npc.c
===================================================================
--- npc.c	(revision 15066)
+++ npc.c	(working copy)
@@ -3114,6 +3136,10 @@
 		map[m].flag.guildlock=state;
 	else if (!strcmpi(w3,"reset"))
 		map[m].flag.reset=state;
+	else if(!strcmpi(w3,"team"))
+		map[m].flag.team=state;
 	else
 		ShowError("npc_parse_mapflag: unrecognized mapflag '%s' (file '%s', line '%d').\n", w3, filepath, strline(buffer,start-buffer));
 
Index: pc.h
===================================================================
--- pc.h	(revision 15066)
+++ pc.h	(working copy)
@@ -137,6 +137,9 @@
 		unsigned short autolootid; // [Zephyrus]
 		unsigned short autobonus; //flag to indicate if an autobonus is activated. [Inkfish]
 		struct guild *gmaster_flag;
+		unsigned int team :6;
 	} state;
 	struct {
 		unsigned char no_weapon_damage, no_magic_damage, no_misc_damage;
Index: script.c
===================================================================
--- script.c	(revision 15066)
+++ script.c	(working copy)
@@ -357,7 +358,9 @@
 	MF_MONSTER_NOTELEPORT,
 	MF_PVP_NOCALCRANK,	//50
 	MF_BATTLEGROUND,
-	MF_RESET
+	MF_RESET,
+	MF_TEAM
 };
 
 const char* script_op2name(int op)
@@ -14963,7 +14966,86 @@
 	return 0;
 }
 
+BUILDIN_FUNC(team)
+{
+	struct map_session_data* sd;
+	const char * name;
+	
+	name = script_getstr(st,2);
+	if(sd = map_nick2sd(name))
+		sd->state.team = (script_getnum(st,3) > 5?0:script_getnum(st,3));
+
+	return 0;
+}
+	
 // declarations that were supposed to be exported from npc_chat.c
 #ifdef PCRE_SUPPORT
 BUILDIN_FUNC(defpattern);
@@ -14976,6 +15058,11 @@
 /// for an explanation on args, see add_buildin_func
 struct script_function buildin_func[] = {
 	// NPC interaction
+	BUILDIN_DEF(team,"si"),
 	BUILDIN_DEF(mes,"s"),
 	BUILDIN_DEF(next,""),
 	BUILDIN_DEF(close,""),
